plugins/data_coverage/src/openjdk/jcov/data/arguments/runtime/Collect.java

```diff
@@ -28,29 +28,107 @@
+            "coverage";
+    /**
+     * Specifies where to load previously collected data from. A non-empty value of this property will
+     * make Collect class to load the data on class loading.
+     */
+    public static final String COVERAGE_IN = Env.JCOV_DATA_ENV_PREFIX + ARGUMENTS_PREFIX +
+            "coverage.in";
+
+    /**
+     * Name of a property containing a class name of a class of type <code>Function<Object, String></code> which will
+     * be used during the serialization. <code>Object::toString</code> is used by default.
+     */
+    public static final String SERIALIZER = Env.JCOV_DATA_ENV_PREFIX +
+            Collect.ARGUMENTS_PREFIX + "serializer";
+
+    static volatile Coverage data;
+    private volatile static Serializer serializer;
 
-    static{
+    static {
+        if (!Env.getStringEnv(COVERAGE_IN, "").isEmpty()) {
+            try {
+                Path coverageFile = Env.getPathEnv(COVERAGE_IN, null);
+                System.out.println("Loading data coverage from " + coverageFile);
+                data = Coverage.read(coverageFile);
+            } catch (IOException e) {
+                throw new UncheckedIOException(e);
+            }
+        } else data = new Coverage();
         try {
-            data = Coverage.readTemplate(Env.getPathEnv(TEMPLATE_FILE, Paths.get("template.lst")));
-        } catch (IOException e) {
-            throw new UncheckedIOException(e);
+            serializer = wrap(Env.getSPIEnv(SERIALIZER, Object::toString));
+        } catch (ClassNotFoundException|NoSuchMethodException|IllegalAccessException|InvocationTargetException|InstantiationException e) {
+            throw new RuntimeException(e);
         }
     }
 
+    private static Serializer wrap(Function<Object, String> function) {
+        if(function instanceof Serializer)
+            return (Serializer) function;
+        else
+            return new Saver.NoRuntimeSerializer(function);
+    }
+
     public static synchronized void collect(String owner, String name, String desc, Object... params) {
 //        keep these lines, it is useful for debugging in hard cases
 //        System.out.printf("%s.%s%s: %s\n", owner, name, desc, (params == null) ? "null" :
-//                Arrays.stream(params).map(Object::getClass).map(Class::getName).collect(joining(",")));
-        data.get(owner, name + desc).add(Arrays.asList(params));
+//                Arrays.stream(params).map(Object::getClass).map(Class::getName)
+//                        .collect(java.util.stream.Collectors.joining(",")));
+//        System.out.println(Arrays.stream(params).map(Objects::toString)
+//                .collect(Collectors.joining(",")));
+        data.add(owner, name + desc, Arrays.stream(params).map(serializer::apply)
+                .collect(Collectors.toList()));
+    }
+
+    static int countParams(String desc) {
+        if(!desc.startsWith("(")) throw new IllegalArgumentException("Not a method descriptor: " + desc);
```

Comment:
I'm not sure. But this parser should be updated with MethodTypeDesc.ofDescriptor && parameterCount() That will support all changes at method descriptors like Q references.

Reply:
Yes, as a separate enhancement.